python/ray/serve/backend_state.py

```diff
@@ -1039,3 +1155,30 @@ def update(self) -> bool:
                 self._notify_replica_handles_changed(tag)
                 for tag in transitioned_backend_tags
             ]
+
+        # After observe & shuffle is done with replicas sitting at new state,
+        # determine which deployment goals succeeded or failed.
+        (complete_goal_ids, failed_goal_ids) = self._check_completed_goals()
+
+        for backend_tag, goal_id in complete_goal_ids:
+            self._replica_failed_to_start_counter[backend_tag] = -1
+            self._goal_manager.complete_goal(goal_id)
+        for backend_tag, goal_id in failed_goal_ids:
+            self._replica_failed_to_start_counter[backend_tag] = -1
+            self._goal_manager.complete_goal(
+                goal_id,
+                # TODO:(jiaodong) Do we want to surface original constructor
+                # exception from other actor instead ? Might need controller
+                # level global exception handler
+                RuntimeError(f"Deployment with goal_id {goal_id} "
+                             f"failed to start. See worker exception above."))
+
+            if backend_tag in self._prev_backend_info:
+                logger.error("Reverting backend {backend_tag} to previous "
+                             f"backend info. {self._prev_backend_info}")
```

Comment:
I think mostly because we don't have _repr_ for BackendInfo as i was expecting it will somehow print high-level description of the deployment setup .. but version number makes more sense for sure.